File: itof.ml
(let rec sub_itof x n =
   (*int * float -> float を足し算でやる、nが小さいから問題ない *)
  if n=1 then
    x
  else 
    sub_itof (x+8388608.0) (n-1)
);

(let rec itof i =
   let ai = if i < 0 then -i else i in
   if ai < 8388608 then
     let fi = i2f (ai + 8388608) in
     let ans = fi - 8388608.0 in
     if i < 0 then
       -ans
     else 
       ans
   
   else
     let n = (ai / 8388608) in
     let m = ai%8388608 in
     let ans = (sub_itof 8388608.0 n) + itof(m) in (* 暗黙のキャスト *)
     if i < 0 then
       -ans
     else
       ans
);

File: test/itof/itof.cpp
#include <iostream>
#include <cmath>
#include <cstdio>

using namespace std;

float itof(int i){
  int ai,ans=0;
  if(i<0) ai = -i;
  else ai = i;

  if(i<8388608){
    float fofi = ai + 8388608; //8388608を整数+で足して足されたものをfloatとして解釈
    ans = fofi - 8388608.0;　// fsub
  }
  if(i>=8388608){
    int n= ai/8388608, m = ai%8388608;
    for(int i=0;i<n;i++){
      ans += 8388608.0;
    }
    ans += itof(m);
  }

  if(i<0) return -ans;
  else return ans;
}


int main(){

  for(int i=0;i<100;i++){
    int s = 111111111;
    cout << s+i << " : ";
    printf("%8f \n" , itof(s+i));
}

  return 0;
}
